pkg/metrics/bpf.go

```diff
@@ -36,55 +56,89 @@ func newbpfCollector() *bpfCollector {
 }
 
 func (s *bpfCollector) Describe(ch chan<- *prometheus.Desc) {
-	ch <- s.bpfMapsMemory
-	ch <- s.bpfProgMemory
+	prometheus.DescribeByCollect(s, ch)
 }
 
 type memoryEntry struct {
 	BytesMemlock uint64 `json:"bytes_memlock"`
 }
 
-func getMemoryUsage(typ string) (uint64, error) {
+func getBPFUsage(typ string) (bpfUsage, error) {
 	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
 	defer cancel()
 	cmd := exec.CommandContext(ctx, "bpftool", "-j", typ, "show")
 	out, err := cmd.Output()
 	if err != nil {
-		return 0, fmt.Errorf("unable to get bpftool output: %w", err)
+		return bpfUsage{}, fmt.Errorf("unable to get bpftool output: %w", err)
 	}
 
 	var memoryEntries []memoryEntry
 	err = json.Unmarshal(out, &memoryEntries)
 	if err != nil {
-		return 0, fmt.Errorf("unable to unmarshal bpftool output: %w", err)
+		return bpfUsage{}, fmt.Errorf("unable to unmarshal bpftool output: %w", err)
 	}
 	var totalMem uint64
 	for _, entry := range memoryEntries {
 		totalMem += entry.BytesMemlock
 	}
-	return totalMem, nil
+
+	return bpfUsage{
+		count:                 uint64(len(memoryEntries)),
+		virtualMemoryMaxBytes: totalMem,
+	}, nil
 }
 
 func (s *bpfCollector) Collect(ch chan<- prometheus.Metric) {
-	mapMem, err := getMemoryUsage("map")
-	if err != nil {
-		logrus.WithError(err).Error("Error while getting BPF maps memory usage")
-	} else {
-		ch <- prometheus.MustNewConstMetric(
-			s.bpfMapsMemory,
-			prometheus.GaugeValue,
-			float64(mapMem),
-		)
+	type bpfUsageResults struct {
+		maps     bpfUsage
+		programs bpfUsage
 	}
 
-	progMem, err := getMemoryUsage("prog")
-	if err != nil {
-		logrus.WithError(err).Error("Error while getting BPF progs memory usage")
-	} else {
-		ch <- prometheus.MustNewConstMetric(
-			s.bpfProgMemory,
-			prometheus.GaugeValue,
-			float64(progMem),
+	// Avoid querying BPF multiple times concurrently, if it happens, additional callers will wait for the
+	// first one to finish and reuse its resulting values.
+	results, err, _ := s.sfg.Do("collect", func() (interface{}, error) {
+		var (
+			results = bpfUsageResults{}
+			err     error
 		)
+
+		if results.maps, err = getBPFUsage("map"); err != nil {
+			return results, err
+		}
+
+		if results.programs, err = getBPFUsage("prog"); err != nil {
+			return results, err
+		}
+
+		return results, nil
+	})
+
+	if err != nil {
+		logrus.WithError(err).Error("retrieving BPF maps & programs usage")
+		return
 	}
+
+	ch <- prometheus.MustNewConstMetric(
+		s.bpfMapsCount,
+		prometheus.GaugeValue,
+		float64(results.(bpfUsageResults).maps.count),
```

Comment:
nit: we could do this type assertion just once to make thing less verbose.

Reply:
definitely, I've updated this implementation 👍